File: backend/src/api/routes/knowledge.py
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from src.api.deps import get_db
from src.schemas.knowledge import (
    KnowledgeCategoryCreate, KnowledgeCategoryUpdate, KnowledgeCategoryRead,
    KnowledgeTagCreate, KnowledgeTagUpdate, KnowledgeTagRead,
    KnowledgeConceptCreate, KnowledgeConceptUpdate, KnowledgeConceptRead, KnowledgeConceptDetail,
    KnowledgeRelationshipCreate, KnowledgeRelationshipUpdate, KnowledgeRelationshipRead,
    KnowledgeExampleCreate, KnowledgeExampleUpdate, KnowledgeExampleRead,
    KnowledgeReferenceCreate, KnowledgeReferenceUpdate, KnowledgeReferenceRead,
    KnowledgeSourceCreate, KnowledgeSourceUpdate, KnowledgeSourceRead,
    KnowledgeChunkRead, KnowledgeRevisionRead, KnowledgeRevisionCreate,
    KnowledgeSearchResult, KnowledgeExplorerResponse, KnowledgeExplorerNode, KnowledgeExplorerEdge,
)
from src.crud import knowledge as crud

router = APIRouter()


# --- Categories ---

@router.get("/categories", response_model=list[KnowledgeCategoryRead])
def read_categories(skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000), db: Session = Depends(get_db)):
    return crud.get_categories(db, skip=skip, limit=limit)

@router.get("/categories/{id}", response_model=KnowledgeCategoryRead)
def read_category(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_category(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    return obj

@router.get("/categories/slug/{slug}", response_model=KnowledgeCategoryRead)
def read_category_by_slug(slug: str, db: Session = Depends(get_db)):
    obj = crud.get_category_by_slug(db, slug)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    return obj

@router.post("/categories", response_model=KnowledgeCategoryRead, status_code=status.HTTP_201_CREATED)
def create_category(obj_in: KnowledgeCategoryCreate, db: Session = Depends(get_db)):
    return crud.create_category(db, obj_in=obj_in)

@router.put("/categories/{id}", response_model=KnowledgeCategoryRead)
def update_category(id: UUID, obj_in: KnowledgeCategoryUpdate, db: Session = Depends(get_db)):
    obj = crud.get_category(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    return crud.update_category(db, db_obj=obj, obj_in=obj_in)

@router.delete("/categories/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_category(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_category(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    crud.remove_category(db, id=id)


# --- Tags ---

@router.get("/tags", response_model=list[KnowledgeTagRead])
def read_tags(skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000), db: Session = Depends(get_db)):
    return crud.get_tags(db, skip=skip, limit=limit)

@router.get("/tags/{id}", response_model=KnowledgeTagRead)
def read_tag(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_tag(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    return obj

@router.post("/tags", response_model=KnowledgeTagRead, status_code=status.HTTP_201_CREATED)
def create_tag(obj_in: KnowledgeTagCreate, db: Session = Depends(get_db)):
    return crud.create_tag(db, obj_in=obj_in)

@router.put("/tags/{id}", response_model=KnowledgeTagRead)
def update_tag(id: UUID, obj_in: KnowledgeTagUpdate, db: Session = Depends(get_db)):
    obj = crud.get_tag(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    return crud.update_tag(db, db_obj=obj, obj_in=obj_in)

@router.delete("/tags/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_tag(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_tag(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    crud.remove_tag(db, id=id)


# --- Concepts ---

@router.get("/concepts", response_model=list[KnowledgeConceptRead])
def read_concepts(
    skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000),
    category_id: UUID | None = None, db: Session = Depends(get_db),
):
    return crud.get_concepts(db, skip=skip, limit=limit, category_id=category_id)

@router.get("/concepts/{id}", response_model=KnowledgeConceptDetail)
def read_concept(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_concept(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Concept not found")
    return obj

@router.get("/concepts/slug/{category_id}/{slug}", response_model=KnowledgeConceptRead)
def read_concept_by_slug(category_id: UUID, slug: str, db: Session = Depends(get_db)):
    obj = crud.get_concept_by_slug(db, category_id, slug)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Concept not found")
    return obj

@router.post("/concepts", response_model=KnowledgeConceptRead, status_code=status.HTTP_201_CREATED)
def create_concept(obj_in: KnowledgeConceptCreate, db: Session = Depends(get_db)):
    return crud.create_concept(db, obj_in=obj_in)

@router.put("/concepts/{id}", response_model=KnowledgeConceptRead)
def update_concept(id: UUID, obj_in: KnowledgeConceptUpdate, db: Session = Depends(get_db)):
    obj = crud.get_concept(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Concept not found")
    return crud.update_concept(db, db_obj=obj, obj_in=obj_in)

@router.delete("/concepts/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_concept(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_concept(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Concept not found")
    crud.remove_concept(db, id=id)


# --- Relationships ---

@router.get("/relationships", response_model=list[KnowledgeRelationshipRead])
def read_relationships(
    skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000),
    concept_id: UUID | None = None, db: Session = Depends(get_db),
):
    return crud.get_relationships(db, skip=skip, limit=limit, concept_id=concept_id)

@router.get("/relationships/{id}", response_model=KnowledgeRelationshipRead)
def read_relationship(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_relationship(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Relationship not found")
    return obj

@router.post("/relationships", response_model=KnowledgeRelationshipRead, status_code=status.HTTP_201_CREATED)
def create_relationship(obj_in: KnowledgeRelationshipCreate, db: Session = Depends(get_db)):
    return crud.create_relationship(db, obj_in=obj_in)

@router.put("/relationships/{id}", response_model=KnowledgeRelationshipRead)
def update_relationship(id: UUID, obj_in: KnowledgeRelationshipUpdate, db: Session = Depends(get_db)):
    obj = crud.get_relationship(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Relationship not found")
    return crud.update_relationship(db, db_obj=obj, obj_in=obj_in)

@router.delete("/relationships/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_relationship(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_relationship(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Relationship not found")
    crud.remove_relationship(db, id=id)


# --- Examples ---

@router.get("/examples", response_model=list[KnowledgeExampleRead])
def read_examples(
    skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000),
    concept_id: UUID | None = None, db: Session = Depends(get_db),
):
    return crud.get_examples(db, skip=skip, limit=limit, concept_id=concept_id)

@router.get("/examples/{id}", response_model=KnowledgeExampleRead)
def read_example(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_example(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Example not found")
    return obj

@router.post("/examples", response_model=KnowledgeExampleRead, status_code=status.HTTP_201_CREATED)
def create_example(obj_in: KnowledgeExampleCreate, db: Session = Depends(get_db)):
    return crud.create_example(db, obj_in=obj_in)

@router.put("/examples/{id}", response_model=KnowledgeExampleRead)
def update_example(id: UUID, obj_in: KnowledgeExampleUpdate, db: Session = Depends(get_db)):
    obj = crud.get_example(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Example not found")
    return crud.update_example(db, db_obj=obj, obj_in=obj_in)

@router.delete("/examples/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_example(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_example(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Example not found")
    crud.remove_example(db, id=id)


# --- References ---

@router.get("/references", response_model=list[KnowledgeReferenceRead])
def read_references(
    skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000),
    concept_id: UUID | None = None, db: Session = Depends(get_db),
):
    return crud.get_references(db, skip=skip, limit=limit, concept_id=concept_id)

@router.get("/references/{id}", response_model=KnowledgeReferenceRead)
def read_reference(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_reference(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reference not found")
    return obj

@router.post("/references", response_model=KnowledgeReferenceRead, status_code=status.HTTP_201_CREATED)
def create_reference(obj_in: KnowledgeReferenceCreate, db: Session = Depends(get_db)):
    return crud.create_reference(db, obj_in=obj_in)

@router.put("/references/{id}", response_model=KnowledgeReferenceRead)
def update_reference(id: UUID, obj_in: KnowledgeReferenceUpdate, db: Session = Depends(get_db)):
    obj = crud.get_reference(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reference not found")
    return crud.update_reference(db, db_obj=obj, obj_in=obj_in)

@router.delete("/references/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_reference(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_reference(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Reference not found")
    crud.remove_reference(db, id=id)


# --- Sources ---

@router.get("/sources", response_model=list[KnowledgeSourceRead])
def read_sources(skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000), db: Session = Depends(get_db)):
    return crud.get_sources(db, skip=skip, limit=limit)

@router.get("/sources/{id}", response_model=KnowledgeSourceRead)
def read_source(id: UUID, db: Session = Depends(get_db)):
    obj = crud.get_source(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Source not found")
    return obj

@router.post("/sources", response_model=KnowledgeSourceRead, status_code=status.HTTP_201_CREATED)
def create_source(obj_in: KnowledgeSourceCreate, db: Session = Depends(get_db)):
    return crud.create_source(db, obj_in=obj_in)

@router.put("/sources/{id}", response_model=KnowledgeSourceRead)
def update_source(id: UUID, obj_in: KnowledgeSourceUpdate, db: Session = Depends(get_db)):
    obj = crud.get_source(db, id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Source not found")
    return crud.update_source(db, db_obj=obj, obj_in=obj_in)


# --- Search ---

@router.get("/search", response_model=list[KnowledgeSearchResult])
def search_knowledge(
    q: str = Query(..., min_length=1),
    category_id: UUID | None = None,
    tag_ids: str | None = None,
    difficulty: str | None = None,
    status: str | None = None,
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    tag_uuid_list = None
    if tag_ids:
        try:
            tag_uuid_list = [UUID(t) for t in tag_ids.split(",")]
        except ValueError:
            pass
    return crud.search_concepts(
        db, q=q, category_id=category_id, tag_ids=tag_uuid_list,
        difficulty=difficulty, status=status, limit=limit, offset=offset,
    )


# --- Related Concepts ---

@router.get("/concepts/{id}/related", response_model=list[KnowledgeConceptRead])
def read_related_concepts(id: UUID, limit: int = Query(20, ge=1, le=100), db: Session = Depends(get_db)):
    return crud.get_related_concepts(db, id, limit=limit)


# --- Explorer (graph data) ---

@router.get("/explorer", response_model=KnowledgeExplorerResponse)
def read_explorer(concept_id: UUID | None = None, db: Session = Depends(get_db)):
    nodes, edges = crud.get_explorer_data(db, concept_id)
    return KnowledgeExplorerResponse(nodes=nodes, edges=edges)


# --- Revisions ---

@router.get("/revisions", response_model=list[KnowledgeRevisionRead])
def read_revisions(
    concept_id: UUID | None = None,
    skip: int = Query(0, ge=0), limit: int = Query(100, ge=1, le=1000),
    db: Session = Depends(get_db),
):
    return crud.get_revisions(db, concept_id=concept_id, skip=skip, limit=limit)

@router.post("/revisions", response_model=KnowledgeRevisionRead, status_code=status.HTTP_201_CREATED)
def create_revision(obj_in: KnowledgeRevisionCreate, db: Session = Depends(get_db)):
    return crud.create_revision(db, obj_in=obj_in)


# --- Dashboard Stats ---

@router.get("/stats")
def read_knowledge_stats(db: Session = Depends(get_db)):
    from sqlalchemy import select, func
    total_categories = db.scalar(select(func.count()).select_from(crud.KnowledgeCategory))
    total_concepts = db.scalar(select(func.count()).select_from(crud.KnowledgeConcept))
    total_relationships = db.scalar(select(func.count()).select_from(crud.KnowledgeRelationship))
    total_examples = db.scalar(select(func.count()).select_from(crud.KnowledgeExample))
    total_references = db.scalar(select(func.count()).select_from(crud.KnowledgeReference))
    return {
        "total_categories": total_categories or 0,
        "total_concepts": total_concepts or 0,
        "total_relationships": total_relationships or 0,
        "total_examples": total_examples or 0,
        "total_references": total_references or 0,
    }

File: frontend/src/api/knowledge.ts
import { supabase } from '../lib/supabase';
import type { KnowledgeCategory, KnowledgeTag, KnowledgeConcept, KnowledgeConceptDetail, KnowledgeRelationship, KnowledgeExample, KnowledgeReference, KnowledgeStats, KnowledgeSearchResult } from './types';

export type { KnowledgeCategory, KnowledgeTag, KnowledgeConcept, KnowledgeConceptDetail, KnowledgeRelationship, KnowledgeExample, KnowledgeReference, KnowledgeStats, KnowledgeSearchResult };

const KNOWLEDGE_TABLES = {
  categories: 'knowledge_category',
  tags: 'knowledge_tag',
  concepts: 'knowledge_concept',
  relationships: 'knowledge_relationship',
  examples: 'knowledge_example',
  references: 'knowledge_reference',
} as const;

export const knowledgeService = {
  getStats: async (): Promise<KnowledgeStats> => {
    const { count: totalCategories } = await supabase.from(KNOWLEDGE_TABLES.categories).select('*', { count: 'exact', head: true });
    const { count: totalConcepts } = await supabase.from(KNOWLEDGE_TABLES.concepts).select('*', { count: 'exact', head: true });
    const { count: totalRelationships } = await supabase.from(KNOWLEDGE_TABLES.relationships).select('*', { count: 'exact', head: true });
    const { count: totalExamples } = await supabase.from(KNOWLEDGE_TABLES.examples).select('*', { count: 'exact', head: true });
    const { count: totalReferences } = await supabase.from(KNOWLEDGE_TABLES.references).select('*', { count: 'exact', head: true });
    return { total_categories: totalCategories ?? 0, total_concepts: totalConcepts ?? 0, total_relationships: totalRelationships ?? 0, total_examples: totalExamples ?? 0, total_references: totalReferences ?? 0 };
  },

  getCategories: async (): Promise<KnowledgeCategory[]> => {
    const { data, error } = await supabase.from(KNOWLEDGE_TABLES.categories).select('*').order('sort_order', { ascending: true }).order('name', { ascending: true });
    if (error) throw error;
    return (data ?? []) as KnowledgeCategory[];
  },
  getCategory: async (id: string): Promise<KnowledgeCategory> => {
    const { data, error } = await supabase.from(KNOWLEDGE_TABLES.categories).select('*').eq('id', id).single();
    if (error) throw error;
    return data as KnowledgeCategory;
  },
  createCategory: async (data: { name: string; slug: string; description?: string; icon?: string; color?: string }): Promise<KnowledgeCategory> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.categories).insert(data).select().single();
    if (error) throw error;
    return row as KnowledgeCategory;
  },
  updateCategory: async (id: string, data: Partial<KnowledgeCategory>): Promise<KnowledgeCategory> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.categories).update(data).eq('id', id).select().single();
    if (error) throw error;
    return row as KnowledgeCategory;
  },
  deleteCategory: async (id: string): Promise<void> => {
    const { error } = await supabase.from(KNOWLEDGE_TABLES.categories).delete().eq('id', id);
    if (error) throw error;
  },

  getTags: async (): Promise<KnowledgeTag[]> => {
    const { data, error } = await supabase.from(KNOWLEDGE_TABLES.tags).select('*').order('name', { ascending: true });
    if (error) throw error;
    return (data ?? []) as KnowledgeTag[];
  },
  createTag: async (data: { name: string; color?: string }): Promise<KnowledgeTag> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.tags).insert(data).select().single();
    if (error) throw error;
    return row as KnowledgeTag;
  },
  deleteTag: async (id: string): Promise<void> => {
    const { error } = await supabase.from(KNOWLEDGE_TABLES.tags).delete().eq('id', id);
    if (error) throw error;
  },

  getConcepts: async (categoryId?: string): Promise<KnowledgeConcept[]> => {
    let query = supabase.from(KNOWLEDGE_TABLES.concepts).select('*, category:category_id(*), tags:knowledge_concept_tag(tag:tag_id(*)))');
    if (categoryId) query = query.eq('category_id', categoryId);
    const { data, error } = await query.order('title', { ascending: true });
    if (error) throw error;
    return (data ?? []) as unknown as KnowledgeConcept[];
  },
  getConcept: async (id: string): Promise<KnowledgeConceptDetail> => {
    const { data, error } = await supabase.from(KNOWLEDGE_TABLES.concepts).select('*').eq('id', id).single();
    if (error) throw error;
    const { data: outgoing } = await supabase.from(KNOWLEDGE_TABLES.relationships).select('*, target_concept:target_concept_id(*)').eq('source_concept_id', id);
    const { data: incoming } = await supabase.from(KNOWLEDGE_TABLES.relationships).select('*, source_concept:source_concept_id(*)').eq('target_concept_id', id);
    const { data: examples } = await supabase.from(KNOWLEDGE_TABLES.examples).select('*').eq('concept_id', id);
    const { data: references } = await supabase.from(KNOWLEDGE_TABLES.references).select('*').eq('concept_id', id);
    return { ...data, relationships_outgoing: outgoing ?? [], relationships_incoming: incoming ?? [], examples: examples ?? [], references: references ?? [], revisions: [] } as KnowledgeConceptDetail;
  },
  getRelatedConcepts: async (_id: string): Promise<KnowledgeConcept[]> => {
    throw new Error('Related concepts requires AI');
  },
  createConcept: async (data: { category_id: string; title: string; slug: string; summary?: string; definition?: string; tag_ids?: string[]; status?: string; difficulty?: string }): Promise<KnowledgeConcept> => {
    const { tag_ids, ...rest } = data;
    const { data: concept, error } = await supabase.from(KNOWLEDGE_TABLES.concepts).insert(rest).select().single();
    if (error) throw error;
    if (tag_ids && tag_ids.length > 0) {
      await supabase.from('knowledge_concept_tag').insert(tag_ids.map((tag_id: string) => ({ concept_id: concept.id, tag_id })));
    }
    return concept as KnowledgeConcept;
  },
  updateConcept: async (id: string, data: Partial<KnowledgeConcept>): Promise<KnowledgeConcept> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.concepts).update(data).eq('id', id).select().single();
    if (error) throw error;
    return row as KnowledgeConcept;
  },
  deleteConcept: async (id: string): Promise<void> => {
    const { error } = await supabase.from(KNOWLEDGE_TABLES.concepts).delete().eq('id', id);
    if (error) throw error;
  },

  getRelationships: async (conceptId?: string): Promise<KnowledgeRelationship[]> => {
    let query = supabase.from(KNOWLEDGE_TABLES.relationships).select('*, source_concept:source_concept_id(*), target_concept:target_concept_id(*)');
    if (conceptId) query = query.or(`source_concept_id.eq.${conceptId},target_concept_id.eq.${conceptId}`);
    const { data, error } = await query;
    if (error) throw error;
    return (data ?? []) as unknown as KnowledgeRelationship[];
  },
  createRelationship: async (data: { source_concept_id: string; target_concept_id: string; relationship_type: string; strength?: number; confidence?: number; description?: string }): Promise<KnowledgeRelationship> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.relationships).insert(data).select().single();
    if (error) throw error;
    return row as KnowledgeRelationship;
  },
  deleteRelationship: async (id: string): Promise<void> => {
    const { error } = await supabase.from(KNOWLEDGE_TABLES.relationships).delete().eq('id', id);
    if (error) throw error;
  },

  getExamples: async (conceptId?: string): Promise<KnowledgeExample[]> => {
    let query = supabase.from(KNOWLEDGE_TABLES.examples).select('*');
    if (conceptId) query = query.eq('concept_id', conceptId);
    const { data, error } = await query.order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as KnowledgeExample[];
  },
  createExample: async (data: { concept_id: string; title: string; pair?: string; timeframe?: string; description?: string }): Promise<KnowledgeExample> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.examples).insert(data).select().single();
    if (error) throw error;
    return row as KnowledgeExample;
  },
  deleteExample: async (id: string): Promise<void> => {
    const { error } = await supabase.from(KNOWLEDGE_TABLES.examples).delete().eq('id', id);
    if (error) throw error;
  },

  getReferences: async (conceptId?: string): Promise<KnowledgeReference[]> => {
    let query = supabase.from(KNOWLEDGE_TABLES.references).select('*');
    if (conceptId) query = query.eq('concept_id', conceptId);
    const { data, error } = await query.order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []) as KnowledgeReference[];
  },
  createReference: async (data: { concept_id: string; source_type: string; title: string; author?: string; url?: string }): Promise<KnowledgeReference> => {
    const { data: row, error } = await supabase.from(KNOWLEDGE_TABLES.references).insert(data).select().single();
    if (error) throw error;
    return row as KnowledgeReference;
  },
  deleteReference: async (id: string): Promise<void> => {
    const { error } = await supabase.from(KNOWLEDGE_TABLES.references).delete().eq('id', id);
    if (error) throw error;
  },

  search: async (q: string, _categoryId?: string): Promise<KnowledgeSearchResult[]> => {
    const { data, error } = await supabase.from(KNOWLEDGE_TABLES.concepts)
      .select('id, title, slug, summary, category:category_id(name), difficulty, status')
      .or(`title.ilike.%${q}%,summary.ilike.%${q}%,definition.ilike.%${q}%`)
      .limit(20);
    if (error) throw error;
    return (data ?? []).map((c: any) => ({ id: c.id, name: c.title, type: 'concept', description: c.summary, category: c.category?.name, tags: c.difficulty ? [c.difficulty] : undefined, relevance: 1 } as KnowledgeSearchResult));
  },

  getExplorer: async (_conceptId?: string): Promise<{ nodes: any[]; edges: any[] }> => {
    const { data: categories } = await supabase.from(KNOWLEDGE_TABLES.categories).select('id, name');
    const { data: concepts } = await supabase.from(KNOWLEDGE_TABLES.concepts).select('id, title, category_id').limit(50);
    const { data: relationships } = await supabase.from(KNOWLEDGE_TABLES.relationships).select('source_concept_id, target_concept_id, relationship_type').limit(100);
    const nodes = [
      ...(categories ?? []).map((c: any) => ({ id: `cat-${c.id}`, label: c.name, type: 'category' })),
      ...(concepts ?? []).map((c: any) => ({ id: `con-${c.id}`, label: c.title, type: 'concept', category_id: c.category_id })),
    ];
    const edges = (relationships ?? []).map((r: any) => ({ source: `con-${r.source_concept_id}`, target: `con-${r.target_concept_id}`, label: r.relationship_type }));
    return { nodes, edges };
  },
};
